File: java/03-bootstrap/src/main/java/com/fResult/bootstrap/customers/services/BaseCustomerService.java
package com.fResult.bootstrap.customers.services;

import com.fResult.bootstrap.customers.Customer;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Objects;
import javax.sql.DataSource;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.PreparedStatementCreator;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.util.Assert;

public class BaseCustomerService implements CustomerService {
  private final RowMapper<Customer> rowMapper =
      (rs, rowNum) -> new Customer(rs.getLong("id"), rs.getString("NAME"));
  private final JdbcTemplate jdbcTemplate;

  public BaseCustomerService(DataSource ds) {
    this.jdbcTemplate = new JdbcTemplate(ds);
  }

  @Override
  public Collection<Customer> save(String... names) {
    final var customers = new ArrayList<Customer>();
    for (final var name : names) {
      final var keyHolder = new GeneratedKeyHolder();

      jdbcTemplate.update(prepareInsertedCustomerStatement(name), keyHolder);

      final var keyHolderKey = Objects.requireNonNull(keyHolder.getKey()).longValue();
      final var customer = findById(keyHolderKey);
      Assert.notNull(name, "The name given must not be null");
      customers.add(customer);
    }

    return customers;
  }

  @Override
  public Customer findById(Long id) {
    final var sql = "SELECT * FROM customers WHERE id = ?";
    return jdbcTemplate.queryForObject(sql, rowMapper, id);
  }

  @Override
  public Collection<Customer> findAll() {
    return jdbcTemplate.query("SELECT * FROM customers", rowMapper);
  }

  private PreparedStatementCreator prepareInsertedCustomerStatement(String name) {
    return connection -> {
      final var ps =
          connection.prepareStatement(
              "INSERT INTO customers (name) VALUES(?)", Statement.RETURN_GENERATED_KEYS);
      ps.setString(1, name);
      return ps;
    };
  }
}

File: kotlin/03-bootstrap/src/main/kotlin/com/fResult/bootstrap/customers/services/BaseCustomerService.kt
package com.fResult.bootstrap.customers.services

import com.fResult.bootstrap.customers.Customer
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.support.GeneratedKeyHolder
import java.sql.Statement
import javax.sql.DataSource

open class BaseCustomerService(ds: DataSource) : CustomerService {
  private val rowMapper = RowMapper { rs, _ -> Customer(rs.getLong("id"), rs.getString("NAME")) }
  private val jdbcTemplate = JdbcTemplate(ds)

  override fun save(vararg names: String): Collection<Customer> {
    val customers = ArrayList<Customer>()
    for (name in names) {
      val keyHolder = GeneratedKeyHolder()

      jdbcTemplate.update({ connection ->
        connection.prepareStatement("INSERT INTO customers (name) VALUES(?)", Statement.RETURN_GENERATED_KEYS)
          .also { it.setString(1, name) }
      }, keyHolder)

      keyHolder.key?.also { key ->
        val keyHolderKey = key.toLong()
        val customer = findById(keyHolderKey)

        customer?.also(customers::add)
      }
    }

    return customers
  }

  override fun findById(id: Long): Customer? {
    val sql = "SELECT * FROM customers WHERE id = ?"
    return jdbcTemplate.queryForObject(sql, rowMapper, id)
  }

  override fun findAll(): Collection<Customer> {
    return jdbcTemplate.query("SELECT * FROM customers", rowMapper)
  }
}
